File: core/org.eclipse.edt.compiler/src/org/eclipse/edt/compiler/internal/core/lookup/ServiceBinder.java
/*******************************************************************************
 * Copyright © 2011, 2013 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * IBM Corporation - initial API and implementation
 *
 *******************************************************************************/
package org.eclipse.edt.compiler.internal.core.lookup;

import org.eclipse.edt.compiler.binding.IRPartBinding;
import org.eclipse.edt.compiler.binding.ServiceBindingCompletor;
import org.eclipse.edt.compiler.core.ast.Service;
import org.eclipse.edt.compiler.internal.core.builder.IProblemRequestor;
import org.eclipse.edt.compiler.internal.core.dependency.IDependencyRequestor;


/**
 * @author winghong
 */

public class ServiceBinder extends FunctionContainerBinder {

	private org.eclipse.edt.mof.egl.Service serviceBinding;
    private IRPartBinding irBinding;
    private Scope fileScope;

    public ServiceBinder(IRPartBinding irBinding, Scope fileScope, IDependencyRequestor dependencyRequestor, IProblemRequestor problemRequestor, ICompilerOptions compilerOptions) {
        super(irBinding.getIrPart(), fileScope, dependencyRequestor, problemRequestor, compilerOptions);
        this.irBinding =irBinding;
        this.serviceBinding = (org.eclipse.edt.mof.egl.Service)irBinding.getIrPart();
        this.fileScope = fileScope;
    }

    public boolean visit(Service service) {
        // First we have to complete the service binding (as a side effect some of the AST nodes are bound)
        service.accept(new ServiceBindingCompletor(fileScope, irBinding, dependencyRequestor, problemRequestor, compilerOptions));

        // The current scope only changes once the initial service binding is complete
        currentScope = new ServiceScope(currentScope, serviceBinding);
        
        preprocessPart(service);

        // We will bind the rest of the service now
        return true;
    }
    
    public void endVisit(Service service) {
		doneVisitingPart();
	}
}

File: runtimes/org.eclipse.edt.runtime.javascript/runtime/eglx/services/ServiceBinder.js
/*******************************************************************************
 * Copyright © 2008, 2012 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * IBM Corporation - initial API and implementation
 *
 *******************************************************************************/
egl.defineClass( 'eglx.services', 'ServiceBinder', {
	
    "constructor" : function() { 
        if (egl.eglx.services.$ServiceBinder) return egl.eglx.services.$ServiceBinder;
		egl.eglx.services.$ServiceBinder=this; 
        this.bindFiles = [];
    }, 
	"addBindFile" : function(/*String*/ eglddName) {
	    if (!this.bindFileExists(eglddName)) {
			this.loadBindFile(eglddName);
	    }
	},
	"bindFileExists" : function(/*String*/ eglddName) {
		var rc = false;
		for(var n=0; n < this.bindFiles && rc == false; n++) {
			if (this.bindFiles.name == eglddName) {
			   rc = true;
			}			
		}
		return rc;
	},
	"loadBindFile" : function(/*String*/ eglddName) {
		try{
	        var bindFileVar = "eze$$BindFile_" + eglddName;
		    var bindFile = egl[bindFileVar]();
		    this.bindFiles.push(bindFile);
		}
		catch(e){
	    	throw egl.eglx.services.createServiceInvocationException("CRRUI3650E", [eglddName]);
		}
	},
	"getBinding" : function(/*String*/ eglddName, /*String*/ bindingKey) {
		this.addBindFile(eglddName);
	    var bindFile = null;
	    for (var n = 0; n < this.bindFiles.length && bindFile == null; n++) {
	        if (this.bindFiles[n].name == eglddName) {
	           bindFile = this.bindFiles[n];
	        }
	    }
	    if (bindFile == null) {
	    	//can not find eglddName
	    	throw egl.eglx.services.createServiceInvocationException("CRRUI3650E", [eglddName]);
	       return null;
	    }
	    
	    bindingKey = bindingKey.toLowerCase();
		var binding = null;
		for (var n=0; n < bindFile.bindings.length && binding == null; n++) {
			if (bindingKey == bindFile.bindings[n].name.toLowerCase()) {
				binding = bindFile.bindings[n];
				break;
			}			
		}
		if(binding == null){
		    for (var n = 0; n <  bindFile.includes.length; n++) {
		    	binding = this.getBinding(bindFile.includes[n], bindingKey);
		    	if ( binding != null ) {
		    		break;
		    	}
		    }
			if (binding == null){
				//can not find binding key
		    	throw egl.eglx.services.createServiceInvocationException("CRRUI3651E", [bindingKey, eglddName]);
			}
		}
		return binding;
	}
});

egl.defineClass( "eglx.services", "BindFile", {
    "constructor" : function(/*String*/ eglddName) { 
		this.name = eglddName; 
        this.bindings = [];
        this.includes = [];
    }
});

egl.defineClass( "eglx.services", "WebBinding", {
    "constructor" : function(/*String*/ _name,
                             /*String*/ _interface,
                             /*String*/ _wsdlLocation,
                             /*String*/ _wsdlService,
                             /*String*/ _wsdlPort,
                             /*String*/ _uri) {
        this.type = "web";
        this.name = _name;
        this._interface = _interface;
        this.wsdlLocation = _wsdlLocation;
        this.wsdlService = _wsdlService;
        this.wsdlPort = _wsdlPort;
        this.uri = _uri;
    }
    
});

egl.defineClass( "eglx.services", "RestBinding", {
    "constructor" : function(/*String*/ _name,
                             /*String*/ _baseURI,
                             /*String*/ _sessionCookieId) {
        this.type = "rest";
        this.name = _name;
        this.baseURI = _baseURI;
        this.sessionCookieId = _sessionCookieId;
    }
    
});

egl.defineClass( "eglx.services", "DedicatedBinding", {
    "constructor" : function(/*String*/ _name) {
        this.type = "dedicated";
        this.name = _name;
    }
    
});
new egl.eglx.services.ServiceBinder();
